Dockerfile

```diff
@@ -85,54 +85,73 @@ EOT
-  make build-secretservice PACKAGE=$PACKAGE DESTDIR=/out BINNAME=docker-credential-secretservice-${TARGETOS}-${TARGETARCH}${TARGETVARIANT}
-  xx-verify /out/docker-credential-secretservice-${TARGETOS}-${TARGETARCH}${TARGETVARIANT}
+  make build-pass build-secretservice PACKAGE=$PACKAGE VERSION=$(cat /tmp/.version) REVISION=$(cat /tmp/.revision) DESTDIR=/out
+  xx-verify /out/docker-credential-pass
+  xx-verify /out/docker-credential-secretservice
 EOT
 
 FROM base AS build-darwin
 ARG PACKAGE
-ARG TARGETOS
-ARG TARGETARCH
-ARG TARGETVARIANT
 RUN --mount=type=bind,target=. \
     --mount=type=cache,target=/root/.cache \
     --mount=type=cache,target=/go/pkg/mod \
-    --mount=type=bind,from=osxcross,src=/osxsdk,target=/xx-sdk <<EOT
+    --mount=type=bind,from=osxcross,src=/osxsdk,target=/xx-sdk \
+    --mount=type=bind,source=/tmp/.version,target=/tmp/.version,from=version \
+    --mount=type=bind,source=/tmp/.revision,target=/tmp/.revision,from=version <<EOT
   set -ex
   xx-go --wrap
   go install std
-  make build-osxkeychain PACKAGE=$PACKAGE DESTDIR=/out BINNAME=docker-credential-osxkeychain-${TARGETARCH}${TARGETVARIANT}
-  xx-verify /out/docker-credential-osxkeychain-${TARGETARCH}${TARGETVARIANT}
-  make build-pass PACKAGE=$PACKAGE DESTDIR=/out BINNAME=docker-credential-pass-${TARGETOS}-${TARGETARCH}${TARGETVARIANT}
-  xx-verify /out/docker-credential-pass-${TARGETOS}-${TARGETARCH}${TARGETVARIANT}
+  make build-osxkeychain build-pass PACKAGE=$PACKAGE VERSION=$(cat /tmp/.version) REVISION=$(cat /tmp/.revision) DESTDIR=/out
+  xx-verify /out/docker-credential-osxkeychain
+  xx-verify /out/docker-credential-pass
 EOT
 
 FROM base AS build-windows
 ARG PACKAGE
-ARG TARGETARCH
-ARG TARGETVARIANT
 RUN --mount=type=bind,target=. \
     --mount=type=cache,target=/root/.cache \
-    --mount=type=cache,target=/go/pkg/mod <<EOT
+    --mount=type=cache,target=/go/pkg/mod \
+    --mount=type=bind,source=/tmp/.version,target=/tmp/.version,from=version \
+    --mount=type=bind,source=/tmp/.revision,target=/tmp/.revision,from=version <<EOT
   set -ex
   xx-go --wrap
-  make build-wincred PACKAGE=$PACKAGE DESTDIR=/out BINNAME=docker-credential-wincred-${TARGETARCH}${TARGETVARIANT}.exe
-  xx-verify /out/docker-credential-wincred-${TARGETARCH}${TARGETVARIANT}.exe
+  make build-wincred PACKAGE=$PACKAGE VERSION=$(cat /tmp/.version) REVISION=$(cat /tmp/.revision) DESTDIR=/out BINNAME=docker-credential-wincred.exe
+  xx-verify /out/docker-credential-wincred.exe
 EOT
 
 FROM build-$TARGETOS AS build
 
 FROM scratch AS binaries
 COPY --from=build /out /
+
+FROM --platform=$BUILDPLATFORM alpine AS releaser
+WORKDIR /work
+ARG TARGETOS
+ARG TARGETARCH
+ARG TARGETVARIANT
+RUN --mount=from=binaries \
+    --mount=type=bind,source=/tmp/.version,target=/tmp/.version,from=version <<EOT
+  set -e
+  mkdir /out
+  for f in *; do
+    cp "$f" "/out/${f%.*}-$(cat /tmp/.version).$TARGETOS-$TARGETARCH$TARGETVARIANT$([ "$TARGETOS" = "windows" ] && echo ".exe")"
```

Comment:
The multiple variables after each other makes it quite hard to read; perhaps use `{}` around them?

```suggestion
    cp "$f" "/out/${f%.*}-$(cat /tmp/.version).${TARGETOS}-${TARGETARCH}${TARGETVARIANT}$([ "$TARGETOS" = "windows" ] && echo ".exe")"
```

That said, there's a lot happening in that one line; I wonder if it would be more readable if some things are split. For example, the version would be the same for every file, so we could just capture that in a variable first. (Same for the file extension?)

```bash
  version="$(cat /tmp/.version)" \
  [ "$TARGETOS" = "windows" ] && file_ext=".exe"
  for f in *; do
    [ -n "${filename##*.}" ] && file_ext=".${filename##*.}"
    cp "$f" "/out/${f%.*}-${version}.${TARGETOS}-${TARGETARCH}${TARGETVARIANT}${file_ext}"
  done
```

It's a pity we need to mangle the name, as the original file already has the proper file extension 🤔. 

```bash
cp docker-credential-wincred.exe /out/docker-credential-wincred-v0.6.4-44-g0d1f2ec.m.windows-amd64.exe
```

Using a directory structure doesn't work for this? (Guess that's not an option if the goal is to publish to GitHub 😞 
Or.. would an alternative be too `.gz` them (keep the original name for the binary, but name the tar with version and platform?)

Reply:
Change a bit to have smth more readable.

> Using a directory structure doesn't work for this? (Guess that's not an option if the goal is to publish to GitHub 😞
> Or.. would an alternative be too `.gz` them (keep the original name for the binary, but name the tar with version and platform?)

Don't think a tarball is necessary. Would also need to produce `.zip` for windows.

I think it's fine like this. We are not going to change that often anyway.